File: bin/tsomeshoptimize.rb
#!ir
$LOAD_PATH.unshift File.expand_path(File.dirname(__FILE__) + "/../lib")
require 'directx'

$LOAD_PATH.unshift File.expand_path(File.dirname(__FILE__) + "/../Debug")
require 'TDCG'
require 'TDCG.NvTriStrip'

require 'unified_position_texcoord_vertex'
require 'unified_position_spec_vertex'
require 'tsoface'

if ARGV.size < 1
  puts "tsomeshoptimize <tso file>"
  exit
end
source_file = ARGV[0]

tso = TDCG::TSOFile.new
tso.load(source_file)

puts "Meshes:"
tso.meshes.each_with_index do |mesh, i|
  puts sprintf("%d %s", i, mesh.name)
end

print "Select mesh (0-#{tso.meshes.size - 1}): "
line = gets
mesh_idx = line.chomp.to_i
mesh_idx = 0
selected_mesh = tso.meshes[mesh_idx]

nodemap = {}
for node in tso.nodes
  nodemap[node.name.to_s] = node
end

def create_faces(mesh)
  faces = []
  for sub in mesh.sub_meshes
    vertices = []
    for a in sub.vertices
      v = UnifiedPositionSpecVertex.new(a, sub)
      vertices.push(v)
    end
    for i in 2...vertices.size
      if i % 2 != 0
        a = vertices[i-2]
        b = vertices[i-0]
        c = vertices[i-1]
      else
        a = vertices[i-2]
        b = vertices[i-1]
        c = vertices[i-0]
      end
      if !a.eql?(b) && !b.eql?(c) && !c.eql?(a)
        f = TSOFace.new(a, b, c)
        faces.push f
      end
    end
  end
  faces
end

WEIGHT_EPSILON = Float::EPSILON # or 1.0e-4

puts "Sub meshes:"
puts "  vertices bone_indices"
puts "  -------- ------------"
for sub in selected_mesh.sub_meshes
  puts sprintf("  %8d %12d", sub.vertices.size, sub.bone_indices.size)
end

print "Set max palettes: "
line = gets
max_palettes = line.chomp.to_i
max_palettes = 16 if max_palettes == 0

def create_vertex(v, bmap)
  a = UnifiedPositionTexcoordVertex.new
  a.position = v.position
  a.skin_weights = System::Array[TDCG::SkinWeight].new(4)
  4.times do |i|
    a.skin_weights[i] =
      if v.skin_weights[i].weight < WEIGHT_EPSILON
        TDCG::SkinWeight.new(0, 0.0)
      else
        TDCG::SkinWeight.new(bmap[v.skin_weights[i].bone_index], v.skin_weights[i].weight)
      end
  end
  a.generate_bone_indices
  a.normal = v.normal
  a.u = v.u
  a.v = v.v
  a
end

def create_sub_meshes(faces, max_palettes)
  faces_1 = faces
  faces_2 = []

  subs = []

  puts "  vertices bone_indices"
  puts "  -------- ------------"

  until faces_1.empty?
    spec = faces_1[0].spec
    bmap = {}
    bone_indices = []
    vmap = {}
    vertices = []
    vert_indices = []

    for f in faces_1
      if f.spec != spec
        faces_2.push(f)
        next
      end

      valid = true
      tmap = {}
      for v in f.vertices
        for sw in v.skin_weights
          next if sw.weight < WEIGHT_EPSILON
          next if bmap[sw.bone_index]
          if bmap.size == max_palettes
            valid = false
            break
          end
          tmap[sw.bone_index] = true
          if bmap.size + tmap.size > max_palettes
            valid = false
            break
          end
        end
      end

      unless valid
        faces_2.push(f)
        next
      end

      for bone_index in tmap.keys
        bmap[bone_index] = bone_indices.size
        bone_indices.push(bone_index)
      end

      for v in f.vertices
        a = create_vertex(v, bmap)
        unless vmap[a]
          vmap[a] = vertices.size
          vertices.push(a)
        end
        vert_indices.push(vmap[a])
      end
    end

    # puts "#vert_indices:#{ vert_indices.size }"
    vert_indices_ary = System::Array[System::UInt16].new(vert_indices.size)
    vert_indices.each_with_index do |vidx, i|
      vert_indices_ary[i] = vidx
    end

    optimized_indices = TDCG::NvTriStrip.optimize(vert_indices_ary)
    # puts "#optimized_indices:#{ optimized_indices.size }"

    sub = TDCG::TSOSubMesh.new
    sub.spec = spec

    # puts "#bone_indices:#{ bone_indices.size }"
    bone_indices_ary = System::Array[System::Int32].new(bone_indices.size)
    bone_indices.each_with_index do |bidx, i|
      bone_indices_ary[i] = bidx
    end
    sub.bone_indices = bone_indices_ary

    vertices_ary = System::Array[UnifiedPositionTexcoordVertex].new(optimized_indices.size)
    optimized_indices.each_with_index do |vidx, i|
      vertices_ary[i] = vertices[vidx]
    end
    sub.vertices = vertices_ary

    puts sprintf("  %8d %12d", sub.vertices.size, sub.bone_indices.size)

    subs.push(sub)

    t = faces_1
    faces_1 = faces_2
    faces_2 = t
    t.clear
  end
  subs
end

def rebuild_mesh(mesh, max_palettes)
  faces = create_faces(mesh)
  # puts "#uniq faces:#{ faces.size }"

  subs = create_sub_meshes(faces, max_palettes)
  # puts "#subs:#{ subs.size }"

  subs_ary = System::Array[TDCG::TSOSubMesh].new(subs.size)
  subs.each_with_index do |sub, i|
    subs_ary[i] = sub
  end
  mesh.sub_meshes = subs_ary
end

rebuild_mesh(selected_mesh, max_palettes)
tso.save('out.tso')

File: trunk/blender/tsomeshoptimize.py
#!/usr/bin/env python
import nvtristrip
from tdcg import TSOFile, TSOSubMesh
import sys

if len(sys.argv) < 2:
	print "tsomeshoptimize <tso file>"
	quit()

source_file = sys.argv[1]

tso = TSOFile()
tso.load(source_file)

print "Meshes:"
for i, mesh in enumerate(tso.meshes):
	print i, mesh.name

mesh_idx = 5
selected_mesh = tso.meshes[mesh_idx]

max_palettes = 12

print "Sub meshes:"
print "  vertices bone_indices"
print "  -------- ------------"
for sub in selected_mesh.sub_meshes:
	print "  %8d %12d" % (len(sub.vertices), len(sub.bone_indices))

class UnifiedPositionSpecVertex(object):
	def __init__(self, a, bone_indices, spec):
		self.co = a.co
		self.no = a.no
		self.uv = a.uv
		self.skin_weights = []
		for bone_index, weight in a.skin_weights:
			self.skin_weights.append(( bone_indices[bone_index], weight ))
		self.spec = spec

	def __eq__(self, v):
		return self.co == v.co and self.spec == v.spec

	def __hash__(self):
		return hash(self.co) ^ hash(self.spec)

class UnifiedPositionTexcoordVertex(object):
	WEIGHT_EPSILON = 1.0e-4

	def __init__(self, v, bone_indices_map):
		self.co = v.co
		self.no = v.no
		self.uv = v.uv
		self.skin_weights = []
		for bone_index, weight in v.skin_weights:
			if weight < WEIGHT_EPSILON:
				self.skin_weights.append(( 0, 0.0 ))
			else:
				self.skin_weights.append(( bone_indices_map[bone_index], weight ))

	def __eq__(self, v):
		return self.co == v.co and self.uv == v.uv

	def __hash__(self):
		return hash(self.co) ^ hash(self.uv)

class TSOFace(object):
	def __init__(self, a, b, c):
		self.a = a
		self.b = b
		self.c = c
		self.spec = a.spec
	
	def vertices(self):
		return ( self.a, self.b, self.c )

def create_faces(mesh):
	faces = []
	for sub in mesh.sub_meshes:
		vertices = [ UnifiedPositionSpecVertex(a, sub.bone_indices, sub.spec) for a in sub.vertices ]
		for i in xrange(2, len(vertices)):
			if i % 2 != 0:
				a = vertices[i-2]
				b = vertices[i-0]
				c = vertices[i-1]
			else:
				a = vertices[i-2]
				b = vertices[i-1]
				c = vertices[i-0]
			if a == b or b == c or c == a:
				continue
			faces.append( TSOFace(a, b, c) )

	return faces

WEIGHT_EPSILON = 1.0e-4

def create_sub_meshes(faces, max_palettes):
	faces_1 = faces
	faces_2 = []

	subs = []

	print "  vertices bone_indices"
	print "  -------- ------------"

	while len(faces_1) != 0:
		spec = faces_1[0].spec
		bmap = {}
		bone_indices = []
		vmap = {}
		vertices = []
		vert_indices = []

		for f in faces_1:
			if f.spec != spec:
				faces_2.append(f)
				continue

			valid = True
			bset = set()
			for v in f.vertices():
				for bone_index, weight in v.skin_weights:
					if weight < WEIGHT_EPSILON:
						continue
					if bone_index in bmap:
						continue
					if len(bmap) == max_palettes:
						valid = False
						break
					bset.add(bone_index)
					if len(bmap) + len(bset) > max_palettes:
						valid = False
						break

			if not valid:
				faces_2.append(f)
				continue

			for bone_index in bset:
				bmap[bone_index] = len(bone_indices)
				bone_indices.append(bone_index)

			for v in f.vertices():
				a = UnifiedPositionTexcoordVertex(v, bmap)
				if a not in vmap:
					vmap[a] = len(vertices)
					vertices.append(a)
				vert_indices.append(vmap[a])

		# print '#vert_indices', len(vert_indices)
		optimized_indices = nvtristrip.optimize(vert_indices)
		# print '#optimized_indices', len(optimized_indices)

		sub = TSOSubMesh()
		sub.spec = spec

		# print '#bone_indices', len(bone_indices)
		sub.bone_indices = bone_indices
		sub.vertices = [ vertices[vidx] for vidx in optimized_indices ]

		print "  %8d %12d" % (len(sub.vertices), len(sub.bone_indices))

		subs.append(sub)

		t = faces_1
		faces_1 = faces_2
		faces_2 = t
		del t[:]

	return subs

def rebuild_mesh(mesh, max_palettes):
	faces = create_faces(mesh)
	print '#uniq faces', len(faces)

	subs = create_sub_meshes(faces, max_palettes)
	print '#subs', len(subs)

	mesh.sub_meshes = subs

rebuild_mesh(selected_mesh, max_palettes)

# vim: set sw=4 ts=4:
